velox/exec/Spiller.cpp

```diff
@@ -547,6 +568,33 @@ void Spiller::finalizeSpill() {
   finalized_ = true;
 }
 
+void Spiller::fillSpillRuns(
+    std::vector<char*>& rows,
+    int32_t numRows,
+    std::vector<uint64_t>& hashes,
+    bool isSinglePartition) {
+  // Calculate hashes for this batch of spill candidates.
+  auto rowSet = folly::Range<char**>(rows.data(), numRows);
+
+  if (!isSinglePartition) {
+    for (auto i = 0; i < container_->keyTypes().size(); ++i) {
+      container_->hash(i, rowSet, i > 0, hashes.data());
+    }
+  }
+
+  // Put each in its run.
+  for (auto i = 0; i < numRows; ++i) {
+    // TODO: consider to cache the hash bits in row container so we only
+    // need to calculate them once.
+    const auto partition = isSinglePartition
+        ? 0
+        : bits_.partition(hashes[i], state_.maxPartitions());
+    VELOX_DCHECK_GE(partition, 0);
+    spillRuns_[partition].rows.push_back(rows[i]);
+    spillRuns_[partition].numBytes += container_->rowSize(rows[i]);
```

Comment:
I guess we could optimize single partition by mem copy from input rows to rows in spillRuns_[partition].rows? We just need the loop to sum up the spill bytes?

Reply:
It seems we could only use std::transform except using for loop as the `spillRuns_[*].rows` type is `std::vector<char*, memory::StlAllocator<char*>>`.

Reply:
Ok, will use `spillRuns_[0].rows = SpillRows(rows.begin(), rows.end(), memory::StlAllocator<char*>{*pool_});`